src/optimization.jl

```diff
@@ -0,0 +1,53 @@
+#=
+For the moment, optimization is delegated to Optim.jl (hardcoded dependency). 
+In order for Optim.jl to use gradients that are returned directly upon function 
+evaluation, one has to use the fix described at: 
+https://github.com/JuliaNLSolvers/Optim.jl/blob/master/docs/src/user/tipsandtricks.md
+
+This is the reason for the cumbersome implementation of optimization with gradients.
+
+Note that by default all particles in the system are assumed optimizable.
+
+IMPORTANT: Note that we always work in cartesian coordinates.
+
+=#
+
+export optimize_geometry
+
+
+""" 
+    By default we work in cartesian coordinaes.
+    Note that internally, when optimizing the cartesian positions, atomic units 
+    are used.
+
+"""
+function Optimization.OptimizationFunction(system::AbstractSystem, calculator; kwargs...)
+    mask = get_optimizable_mask(system) # mask is assumed not to change during optim.
+
+    f = function(x::AbstractVector{<:Real}, p)
+        new_system = update_optimizable_positions(system, x * u"bohr")
+        energy = AtomsCalculators.potential_energy(new_system, calculator; kwargs...)
+        austrip(energy)
+    end
+
+    g! = function(G::AbstractVector{<:Real}, x::AbstractVector{<:Real}, p)
+        new_system = update_optimizable_positions(system, x * u"bohr")
+        energy = AtomsCalculators.potential_energy(new_system, calculator; kwargs...)
+
+        forces = AtomsCalculators.forces(new_system, calculator; kwargs...)
+        # Translate the forces vectors on each particle to a single gradient for the optimization parameter.
+        forces_concat = collect(Iterators.flatten(forces[mask]))
+
+        # NOTE: minus sign since forces are opposite to gradient.
+        G .= - austrip.(forces_concat)
+    end
+    OptimizationFunction(f; grad=g!)
+end
+
+function optimize_geometry(system::AbstractSystem, calculator; solver=Optim.LBFGS(), kwargs...)
+    # Use current system parameters as starting positions.
+    x0 = Vector(austrip.(get_optimizable_positions(system))) # Optim modifies x0 in-place, so need a mutable type.
+    f_opt = OptimizationFunction(system, calculator)
```

Comment:
I wonder if instead of creating a `OptimizationFunction` we should instead create an `OptimizationProblem`, because then we could also take more complex constraints into account more transparently later. Just a thought. Probably does not matter too much either way.

Reply:
I think this is probably a good idea, see my comment above – in SciML (Optimization.jl) I think they have both the `OptimizationFunction` (which is optional since you can pass a function directly as well) and the `OptimizationProblem` objects.

Reply:
I agree that `OptimizationProblem` can be used in the future to implement constraints. Nevertheless, I would argue that we have to keep our definition of `OptimizationFunction` (as compared to passing directly a function to `OptimizationProblem`), since we need it because of the way we compute our gradients (so that some of the algorithms can compute both function value and gradient at the same time).